### ISSUE TITLE
Variable length multi input + custom layer + positive/negative sampling implementation difficulties

### ISSUE THREAD

**Issue by tseckin** [fchollet/keras#2800]
Hi, I am new to Keras and trying to implement this [paper](http://arxiv.org/abs/1602.03609v1). But I am experiencing some difficulties.

For my data, I have questions and answers. These words in my data is replaced by embedded word vectors coming from word2vec, these are pre-trained vectors, no problem. But, I have variable length matrix for each training sample
`[embedding_vector_size x M]` where M is the token count for each question in my data. And answers are similar also.

Now, I tried

```
    model = Graph()

    # Add inputs
    model.add_input('question', input_shape=(None, embedding_size))
    model.add_input('positive_answer', input_shape=(None, embedding_size))
    model.add_input('negative_answer', input_shape=(None, embedding_size))

    model.add_node(Convolution1D(conv_filter_count, context_window_k),
                          name='answer_conv',
                          input='positive_answer')

    model.add_node(Convolution1D(conv_filter_count, context_window_k),
                   name='question_conv',
                   input='question')
```

then I need to implement a layer which learns U matrix, where Q is convolution matrix over questions and A is convolution matrix over its answer. Briefly, what I need to implement as a Layer is,

`tanh(transpose(Q) U A)`

For this purpose, I thought to use merge layer. But it has no build method, thus it doesn't learn a parameter. But I need to make it learn U matrix above. So, I thought to split formula into two and use a Dense layer to learn U. Like, 

_partial_product=U*A_
_full_product = Q*partial_product_
_result = tanh(full_product)_

For this purpose,

```
    model.add_node(Dense(conv_filter_count),
                   name='partial_product',
                   input='question_conv')

    model.add_node(Lambda(get_attention_layer_right),
                   name='product',
                   inputs=['answer_conv', 'partial_product'])

    model.add_node(get_activation(),
                   name="g_matrix",
                   input='product')
```

and

```
def get_attention_layer_right(X):
    answer_conv, question_mult = X.values()
    val = K.dot(answer_conv, question_mult)
    return val

def get_activation():
    model = Sequential()
    model.add(Activation('tanh'))
    return model
```

However I am getting error:

```
assert len(input_shape) == 2
AssertionError
```

How can I implement this layer and how can I solve the error?
Thanks

**Comment by phdowling** [comment]
You will need to mad your data to have the same length, for instance using `keras.preprocessing.sequence.pad_sequences`. If you pad using zeros and use an Embedding layer with mask_zeros=True for your input, the padding should have no negative impact on your model.

**Comment by tseckin** [comment]
Thanks for your answer. As far as I understand, Embedding layer should be first layer and it converts  words' integer corresponding to fixed size vectors. But I want to use already computed embedded words by word2vec and these vectors are also fixed sized. 

My problem is, there are embedded vectors with arbitrary count for each input data. For example, if my question is: "how are you", and embedded vector size for every words is, lets say, L;
then I have **Z=3xL** sized matrix for this example. The variable is 3 here, it is token count of a question. It can change one training example to another training example and I want to convolve that **Z** matrix. (Btw this is simplified, actual [paper](http://arxiv.org/abs/1602.03609v1) explains Z matrix)

**Comment by phdowling** [comment]
You can still load word vectors in the Embedding layers, just specify the weights matrix via the `weights` parameter. If you already have embeddings that are the output of an LSTM or something like that, then the only way to train on those as features directly is to compute them for all samples and use them as input directly (or load them in an Embedding layer and construct your input to the network accordingly).

What I think you want to do is to input variable length sequences into a recurrent layer correct? To do this, construct all sequences, figure out the maximum length of any one sequence, then pad all sequences to that length. Then your input has fixed dimensionality as far as the code is concerned, but the masking in the embedding layer still ensures that the recurrent layer only trains on the actual data, not the padding. Hope this helps.

**Comment by tseckin** [comment]
Yes, I have already computed word vectors coming from Google [word2vec](https://code.google.com/archive/p/word2vec/)  and I tried to use them directly as feature vectors like this:

```
    model.add_input('question', input_shape=(None, embedding_size))
    model.add_input('positive_answer', input_shape=(None, embedding_size))
    model.add_input('negative_answer', input_shape=(None, embedding_size))
```

Actually no, I want to do convolution on these input variable length sequence matrices:

```
   model.add_node(Convolution1D(conv_filter_count, context_window_k),
                   name='question_conv',
                   input='question')
```

However I stucked as I mentioned in my first message. Because I want to implement:
`tanh(transpose(Q) U A)`
and for this purpose I added Dense layer which gives me AssertionError :/
